### DISCUSSION TITLE
Query about coupling function with the PropertyReadFile for both temporal and spatial changes

### DISCUSSION THREAD

**Discussion by DaijunHU** [idaholab/moose#28704] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Dear Moose Users,

I am now trying to simulate cracking at varying temperatures using a phase-field model, in which I want to set that the critical energy release rate Gc is not uniformly distributed in the RVE and changing along with temperature increase.

Firstly I am assigning Gc into the RVE by PropertyReadFile in UserObjects, in the .txt file the G_c is assigned to each element seperately:

```
> [UserObjects]
>   [./Gc_read]
>     type = ElementPropertyReadFile
>     prop_file_name = 'non_linear.txt'
>     nprop = 1
>     read_type = element
>   [../]
> []

```
while in the Materials block, I introduce G_c as (gc_prop_read is added in [AuxVariables] ):
```
  [./Gc_from_file_object]
    type = FileParsedMaterial
    prop_name = 'gc_prop_read'
    read_prop_user_object = Gc_read
  [../] 
```
Then in the Functions block I want to implement the temperature dependence of this G_c:
```
[Functions]
  [./temperature_Gc]
    type = ParsedFunction
    expression = '1.8*(x-300)/356'
  [../]
[]
```
My goal is to let the Gc of each element following this temperature function (temperature is changing at different steps following another function), but still keep the spatial dfiference given by the initial value from the read .txt file.
I tried to use this method in the [Materials] block, but it seems not working:
```
  [./Gc_scale]
    type = CoupledValueFunctionMaterial
    function = temperature_Gc
    v = 'temp'
    prop_name = gc_prop_read
  [../]  
```
Also I tried  DerivativeParsedMaterial, which is also not working well.

Could you please give some suggestions on this issue?

**Comment by GiudGiud** [comment]
Hello

What do you mean by it seems it's not working?

What error do you get with the CoupledVariableFunctionMateeial?
And what about with the DericativeParsedMaterial?

Guillaume

**Reply by DaijunHU** [reply]
Hi,
Thank you for your prompt reply.
Actually I did not meet errors when using them. I can find the Gc are correctly distributed in the RVE as expected. But when temperature changes they did not change together, so I guess the coupling between function and propertyreadfile is not right. I want to know whether I am on a right way  to solve it in [Materials] like this or should use other methods instead.
Thank you!

Best regards,
Daijun

**Reply by DaijunHU** [reply]
Hi Guillaume,

Sorry that I should provide more information for my query.
As mentioned above I made a [Gc_scale] block, where I made the temperature dependence of Gc into a scale factor.
This factor is transfered to an auxvariable called _gc_scale_, while the read gc initial values are stored to another auxvariable _gc_read_.
```
  [./Gc_scale]
    type = CoupledValueFunctionMaterial
    function = temperature_Gc
    v = 'temp'
    prop_name = gc_prop_scale
  [../]  
  [./Gc_from_file_object]
    type = FileParsedMaterial
    prop_name = 'gc_prop_read'
    read_prop_user_object = Gc_read
  [../] 
```
Then my questions goes to how to combine them together to a final _gc_prop_ which is needed for my fracture modeling.
I firstly tried ParsedMaterial like:
```
  [./Gc_temp_map]
    type = ParsedMaterial
    expression = 'gc_prop_scale *gc_prop_read'
    coupled_variables= 'gc_prop_scale  gc_prop_read'
    property_name = gc_prop
  [../] 
```
But it shows that  ERROR: FunctionParser is unable to parse expression: gc_prop_scale * gc_prop_read
Then I tried CoupledValueeFunctionMaterial by adding a function _Gc_calculation_:
```
  [./Gc_calculation]
    type = ParsedFunction
    expression = 'x*y'
  [../]
  [./Gc_temp_map]
    type = CoupledValueFunctionMaterial
    function = Gc_calculation
    v = 'gc_prop_scale  gc_prop_read'
    prop_name = gc_prop
  [../] 
```
It shows no error but not working to give the temperature dependence to Gc. In my opinion this is close and I am also checking the tempearture files used for this method. For my this coupling issue is that the temperature is changing with time, Gc is both temperature dependent and spatially different in each element as assigned.

The DerivativeParsedMaterial is not a right way I think, because currently no derivative is included.

I look forward to your suggestions on using which type for this kind of coupling. If the CoupledValueFunctionMaterial is a correct way, i will check other parts of my input file. Thank you in advance and will keep you updated. 

Best regards,
Daijun

**Reply by GiudGiud** [reply, marked as answer]
hello

```
  [./Gc_temp_map]
    type = ParsedMaterial
    expression = 'gc_prop_scale *gc_prop_read'
    coupled_variables= 'gc_prop_scale  gc_prop_read'
    property_name = gc_prop
  [../] 
```
this is not quite right since gc_prop_scale  and gc_prop_read are not variables, they are material properties
so it should be
```
material_property_names = 'gc_prop_scale  gc_prop_read'
```
instead

**Reply by DaijunHU** [reply]
Dear Guillaume,

Yes that is the point. I have tried and this made it work.
Thank you!
